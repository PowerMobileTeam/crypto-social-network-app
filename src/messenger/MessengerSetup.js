import React, {
  Component
} from 'react';

import {
  Text,
  View,
  TextInput,
  Alert,
  StyleSheet,
} from 'react-native';

import {
  inject,
  observer
} from 'mobx-react/native'

import { CommonStyle } from '../styles/Common';
import { ComponentsStyle } from '../styles/Components';
import Colors from '../styles/Colors';
import Button from '../common/components/Button';
import NavNextButton from '../common/components/NavNextButton';

/**
 * Messenger setup
 */
@inject('messengerList', 'user')
@observer
export default class MessengerSetup extends Component {

  /**
   * password
   * (don't use state to prevent render the component when password change)
   */
  password = '';
  confirm  = '';

  componentWillMount() {
    const { setParams } = this.props.navigation;
    let button;

    if (this.props.user.me.chat && !this.props.rekey) {
      button = (
        <NavNextButton
          onPress={this.unlock}
          title="UNLOCK"
          color={Colors.primary}
        />
      );
    } else {
      button = (
        <NavNextButton
          onPress={this.setup}
          title="SETUP"
          color={Colors.primary}
        />
      );
    }
    setParams({ headerRight: button });
  }

  unlock = () => {
    this.props.messengerList.getCryptoKeys(this.password)
      .then(resp => {
        this.handleOnDone(resp);
      });
  }

  setup = () => {
    if (this.password !== this.confirm) {
      Alert.alert('password and confirmation do not match!');
      return;
    }
    this.props.messengerList.doSetup(this.password)
      .then(resp => {
        this.handleOnDone(resp);
      });
  }

  handleOnDone(resp) {
    if (this.props.onDone) {
      this.props.onDone(resp);
    }
  }

  renderUnlock() {
    const unlocking = this.props.messengerList.unlocking;

    return (
      <View style={[CommonStyle.flexContainer, CommonStyle.padding2x, CommonStyle.backgroundLight]}>
        <View style={{ flexDirection: 'column', alignItems: 'stretch' }}>
          <TextInput
            style={ComponentsStyle.passwordinput}
            editable={true}
            underlineColorAndroid='transparent'
            placeholder='password...'
            secureTextEntry={true}
            onChangeText={(password) => this.password = password}
          />
        </View>

        <View style={{ paddingTop: 32 }}>
          <Text style={styles.infoText}>路 You only need to enter this encryption password once as long as you stay signed in.</Text>
          <Text style={styles.infoText}>路 It is important so that no one other than you and the people you are communicating with can access the content of your messages.</Text>
          <Text style={styles.infoText}>路 By default the content of your messages is fully encrypted. For heightened security you may wish to go to your app settings and turn off push notifications in order to disallow metadata from being tracked.</Text>
          <Text style={styles.infoText}>路 By default the content of your messages is fully encrypted. For heightened security you may wish to go to your app settings and turn off push notifications in order to disallow metadata from being tracked.</Text>
        </View>
      </View>
    )
  }

  renderOnboarding() {
    const unlocking = this.props.messengerList.unlocking;

    const text = this.props.user.me.chat ? 'Changing your encryption password will cause your previous messages to be unreadable':
      'Hey @'+this.props.user.me.name+'! It looks like you haven\'t setup your encrypted chat password yet. We recommend that you use a different password than your account password for added security.'

    return (
      <View style={[CommonStyle.flexContainer, CommonStyle.padding2x, CommonStyle.backgroundLight]}>
        <View style={{ flexDirection: 'column', alignItems: 'stretch' }}>
          <TextInput
            style={ComponentsStyle.passwordinput}
            editable={true}
            underlineColorAndroid='transparent'
            placeholder='password...'
            secureTextEntry={true}
            onChangeText={(password) => this.password = password}
          />
          <TextInput
            style={[ComponentsStyle.passwordinput, CommonStyle.marginTop2x]}
            editable={true}
            underlineColorAndroid='transparent'
            placeholder='confirm password...'
            secureTextEntry={true}
            onChangeText={(password) => this.confirm = password}
          />
        </View>

        <View style={{ paddingTop: 16 }}>
          <Text style={styles.infoText}>{text}</Text>
        </View>
      </View>
    );
  }

  /**
   * Render
   */
  render() {

    if (this.props.user.me.chat && !this.props.rekey) {
      return this.renderUnlock();
    } else {
      return this.renderOnboarding();
    }
  }
}

const styles = StyleSheet.create({
	infoText: {
    marginBottom: 16,
    color: '#BBB',
  },
});
