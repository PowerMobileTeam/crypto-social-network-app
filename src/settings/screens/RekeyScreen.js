import React, {
  Component,
} from 'react';
import {Alert} from 'react-native';

import MessengerSetup from '../../messenger/MessengerSetup';

/**
 * Messenger re key screen
 */
export default class RekeyScreen extends Component {

  static navigationOptions = ({ navigation }) => {
    return {
      headerRight: navigation.state.params && navigation.state.params.headerRight,
    }
  };

  /**
   * On done
   */
  onDone = () => {
    Alert.alert('Keys regenerated successfully!');
    this.props.navigation.goBack();
  }

  /**
   * Render
   */
  render() {
    return <MessengerSetup rekey={true} onDone={this.onDone} navigation={this.props.navigation}/>
  }
}