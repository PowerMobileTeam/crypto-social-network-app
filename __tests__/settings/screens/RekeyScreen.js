import 'react-native';
import React from 'react';
import { shallow } from 'enzyme';

import RekeyScreen from '../../../src/settings/screens/RekeyScreen';
import MessengerSetup from '../../../src/messenger/MessengerSetup';

jest.mock('../../../src/messenger/MessengerSetup', () => 'MessengerSetup');

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';

describe('Rekey screen component', () => {
  it('should renders correctly', () => {

    const navigation = {};

    const rendered = renderer.create(
      <RekeyScreen navigation={navigation}/>
    ).toJSON();

    expect(rendered).toMatchSnapshot();
  });

  it('should go back on done', () => {

    const navigation = {goBack: jest.fn()};

    const wrapper = shallow(
      <RekeyScreen navigation={navigation}/>
    )

    // call method
    wrapper.instance().onDone();

    expect(navigation.goBack).toBeCalled();
  });
});
