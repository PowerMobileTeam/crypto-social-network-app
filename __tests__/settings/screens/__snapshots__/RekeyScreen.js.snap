// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rekey screen component should renders correctly 1`] = `
<MessengerSetup
  navigation={Object {}}
  onDone={[Function]}
  rekey={true}
/>
`;
